README.md
[![Build Status](https://api.travis-ci.org/pjfl/p5-app-doh.svg?branch=master)](https://travis-ci.org/pjfl/p5-app-doh)


**Doh** is a documentation generator that uses a simple folder structure and
[Markdown](http://daringfireball.net/projects/markdown/) files to create
custom documentation on the fly. It helps you create great looking
documentation in a developer friendly way.

## Features

* 100% Mobile responsive
* Supports an extended flavour of multi-markdown
* Automatically created home / landing page
* Automatic syntax highlighting
* Automatically generated navigation
* 4 Built-in colour themes or roll your own
* Functional, flat design style
* Shareable / linkable SEO friendly URLs
* Bootstrap classes on the markup
* No build step
* Git / SVN friendly
* Google analytics
* Optional code float layout
* CodeMirror markdown editor and file management
* Multi-language support
* Static site generator
* Text searching

## Acknowledgements

I saw [Daux.io](https://github.com/justinwalsh/daux.io) on Github and it said
"Fork Me" so I did.

JavaScript libraries hosted by [CloudFlare](https://www.cloudflare.com/).

## Differences

I exchanged the PHP for modern [Perl](http://www.perl.org),
[Moo](https://metacpan.org/module/Moo),
[Plack](https://metacpan.org/module/Plack), and
[Web::Simple](https://metacpan.org/module/Web::Simple). Not a fan of Bootstrap
so that went also. Prefer class based inheritance when JavaScript programming
hence [Mootools](http://mootools.net/) not jQuery.

## Installation

The [App-Doh repository](http://github.com/pjfl/p5-app-doh) contains meta data
that lists the CPAN modules used by the application. Modern Perl CPAN
distribution installers (like
[App::cpanminus](https://metacpan.org/module/App::cpanminus)) use this
information to install the required dependencies when this application is
installed.

**Requirements:**

* Perl 5.12.0 or above
* Git - to install **App::Doh** from Github

To find out if Perl is installed and which version; at a shell prompt type

```shell
   perl -v
```
To find out if Git is installed, type

```shell
   git --version
```

If you don't already have it, bootstrap
[App::cpanminus](https://metacpan.org/module/App::cpanminus) with:

```shell
   curl -L http://cpanmin.us | perl - --sudo App::cpanminus
```

Then install [local::lib](https://metacpan.org/module/local::lib) with:

```shell
   cpanm --notest --local-lib=~/App-Doh local::lib && \
      eval $(perl -I ~/App-Doh/lib/perl5/ -Mlocal::lib=~/App-Doh)
```

The second statement sets environment variables to include the local Perl
library. You can append the output of the perl command to your shell startup if
you want to make it permanent. Without the correct environment settings Perl
will not be able to find the installed dependencies and the following will
fail, badly.

Install **App-Doh** with:

```shell
   cpanm --notest git://github.com/pjfl/p5-app-doh.git
```

Although this is a *simple* application it is composed of many CPAN
distributions and, depending on how many of them are already available,
installation may take a while to complete. The flip side is that there are no
external dependencies like Node.js or Grunt to install. Anyway you are advised
to seek out sustenance whilst you wait for ack-2.12 to run it's tests.  At the
risk of installing broken modules (they are only going into a local library)
you can skip the tests by running `cpanm` with the `--notest` option.

If that fails run it again with the --force option

```shell
   cpanm --force git:...
```

By default the development server will run at:
[http://localhost:5000](http://localhost:5000) and can be started in the
foreground with:

```shell
   cd App-Doh
   plackup bin/doh-server
```

Users must authenticate against the `users.json` file under the document
root. The default user is `admin` password `admin123`. You should change that
via the user profile editor which is the stud that appears on the right of the
header after you have logged in. To start the production server in the
background listening on the default port 8085 use:

```shell
   doh-daemon start
```

The `doh-daemon` program provides normal SysV init script semantics.
Additionally the daemon program will write an init script to standard output in
response to the command:

```shell
   doh-daemon get_init_file
```

## Folders

The generator will look for folders in the `var/root/docs` folder. Add your
folders inside the `var/root/docs` folder. This project contains some example
folders and files to get you started.

The first folder under `var/root/docs` should be a language code. This defaults
to `en` for English. Even if you only plan to use one language you will still
need to create your files and folders underneath the language folder.

You can nest folders any number of levels to get the exact structure you want.
The folder structure will be converted to the nested navigation.

## Files

The generator will look for Markdown `*.md` or `*.mkdn` files inside
the `var/root/docs` folder and any of the subfolders within `var/root/docs`.

You must use either the `.md` or the `.mkdn` file extension for your
files. Also, you must use underscores instead of spaces. Here are some
example file names and what they will be converted to:

**Good:**

* 01_Getting_Started.md = Getting Started
* API_Calls.md = API Calls
* 200_Something_Else-Cool.md = Something Else-Cool

**Bad:**

* File Name With Space.md = FAIL

The generator will also look for files with the extensions; `.pl`, `.pm`, and
`.pod`. It will create markup from Perl's POD contained in these files.

Text files with `.txt` extension will be rendered wrapped in 'pre' tags.

## Sorting

To sort your files and folders in a specific way, you can prefix them with a
number and underscore, e.g. `var/root/docs/en/01_Hello_World.md` and
`var/root/docs/en/05_Features.md`. This will list *Hello World* before
*Features*, overriding the deafult alpha-numeric sorting. The language code,
numbers, and file extensions will be stripped out of the navigation links and
URIs.

## Landing page

If you want to create a beautiful landing page for your project,
simply create a `index.md` file in the root of the `var/root/docs/en`
folder. This file will then be used to create a landing page. You can
also add a description and image to this page using the config file like
this:

```json
   {
      "title": "Doh!",
      "description": "An Easiest Way To Document Your Project",
      "brand": "app.png"
   }
```

## Configuration

To customize the look and feel of your documentation, you can create a
`app-doh.json` file in the of the `var/root/docs` folder. The `app-doh.json`
file is a simple JSON object that you can use to change some of the
basic settings of the documentation. The
[config help](/help/App::Doh::Config) page details the available
configuration attributes

### Title:
Change the title bar in the docs

```json
   {
      "title": "Doh!"
   }
```

### Themes:
We have 4 built-in colour themes. To use one of the themes, just
set the `theme` option to one of the following:

* blue
* green
* navy
* red

You can set the theme option by appending '?theme=&lt;colour&gt;' to any URI.
The value will persist once it has been set.

```json
   {
      "theme": "blue"
   }
```

### Custom Theme:
To create a custom color scheme, set the `theme` property to `custom`
and then define the required colors. Copy the following configuration
to get started:

```json
   {
      "theme": "custom",
      "colors": {
         "sidebar-background": "#f7f7f7",
         "sidebar-hover": "#c5c5cb",
         "lines": "#e7e7e9",
         "dark": "#3f4657",
         "light": "#82becd",
         "text": "#2d2d2d",
         "syntax-string": "#022e99",
         "syntax-comment": "#84989b",
         "syntax-number": "#2f9b92",
         "syntax-label": "#840d7a"
      }
   }
```

### Code Floating:
By deafult your code blocks will be floated to a column on the right
side of your content. To disable this feature, set the `float`
property to `false`, or add `?float=0` to any URI to store the choice
in the preferences. Add `?float=1` to any URI to re-enable the
feature

```json
   {
      "float": false
   }
```

### Ignore:
The config attribute `no_index` is a list of filename patterns to ignore
when indexing the document tree

```json
   [ '.git', '.htpasswd', '.svn', 'app-doh.json' ]
```

### Locales:
The list of locales (languages) supported

```json
   {
      "locales": [ "en", "de" ]
   }
```

### Github Repo:
Add a 'Fork me on Github' ribbon.

```json
   {
      "repo_url": "https://github.com/pjfl/p5-app-doh.git"
   }
```

### Twitter:
Include twitter follow buttons in the sidebar.

```json
   {
      "twitter": [ "perl", "web-simple" ]
   }
```

### Links:
Include custom links in the sidebar.

```json
   {
      "links": {
         "Download": "https://github.com/pjfl/p5-app-doh/archive/master.zip",
         "Github Repo": "https://github.com/pjfl/p5-app-doh.git",
         "Issues/Wishlist": "http://github.com/pjfl/p5-app-doh/issues"
      }
   }
```

### Google Analytics:
This will embed the Google Analytics tracking code.

```json
   {
      "analytics": "UA-XXXXXXXXX-XX"
   }
```

## Toggling Code Blocks
The select menu in the left column of the documentation template allows
for the selection of different code block display modes. Code blocks
can be displayed in the page float view, the can be displayed inline, and
they can be hidden

## Multi-language Support

Use your browsers preferences to set the `Accept-Language` header in the
request to the desired language. If available it will be served in preference
to the default language, English.

The national flag on the right of the title bar indicates the language of the
file being displayed or edited. Adding `?use_flags=0` to a URI will turn the
flags off and display a language code instead.

Text within the application can be translated into other languages by
creating GNU Gettext portable object files. The path is
`var/locale/<language_code>/LC_MESSAGES` and the two file names are
`default.po` and `app_doh.po`. Messages in the `default.po` are common to all
programs in the application. Creating English translation files allows the
terse error messages to be replaced with more user friendly ones.

### Directory structure:

```
   ├── docs/
   │   ├── en/
   │   │   ├── index.md
   │   │   ├── 00_Getting_Started.md
   │   │   ├── 01_Examples/
   │   │   │   ├── 01_GitHub_Flavored_Markdown.md
   │   │   │   ├── 05_Code_Highlighting.md
   │   │   ├── 05_More_Examples/
   │   │   │   ├── Hello_World.md
   │   │   │   ├── 05_Code_Highlighting.md
   │   ├── de/
   │   │   ├── index.md
   │   │   ├── 00_Getting_Started.md
   │   │   ├── 01_Examples/
   │   │   │   ├── 01_GitHub_Flavored_Markdown.md
   │   │   │   ├── 05_Code_Highlighting.md
   │   │   ├── 05_More_Examples/
   │   │   │   ├── Hello_World.md
   │   │   │   ├── 05_Code_Highlighting.md
```

## Markdown Editor
<img alt="[File editor]" class="app-thumbnail"
     src="[% links.images %]app-editor.png"/>

The markdown in the document tree is editable via the web browser. The
`Edit` link is in the left navigation column after the code blocks
selector. The `Save` button at the bottom of the text editor makes the
changes permanent.

The `Create` link displays a modal dialog. Enter the path to the file you want
to create. The path should be relative to the language sub-directory, it
should not begin with a leading slash. Do include numeric prefixes for sorting
purposes, use slashes to delineate directories and files, spaces will be
replaced with underscores, and the file extension is optional. Any new
directories in the file path will be automatically created.

The `Delete` link will prompt for confirmation before deleting the markdown
file. Empty directories will be removed.

The `Upload` link lets you select a file to upload. The uploaded file appears
in the `var/root/docs/assets` directory. It can be referenced from a template
using `[\% links.assets %\]your-asset.jpg`, or from markdown as
`[\%links.assets%\]your-asset.jpg` (no spaces).

## Generating CSS from Less

Only the green theme is precompiled, if you want the other colour themes then
install Node, Lessc, and CSS::LESS. As root:

```shell
   apt-get update
   apt-get install git-core curl build-essential openssl libssl-dev
   git clone https://github.com/joyent/node.git
   cd node
   ./configure --openssl-libpath=/usr/lib/ssl
   make
   make install
   # Node installed with broken permissions which I had to fix at this point
   npm install -g less
```

Now you should have a working `lessc` command. Create the other CSS files with
(not as root):

```shell
   cpanm --force CSS::LESS
   doh-cli make_css
```

## Generating a set of static files

This command creates a static HTML version of the documentation in the
`var/root/static` directory

```shell
   doh-cli make_static
```

## Support

There is a link on the right side of the title bar that displays the Perl
documentation for [App::Doh](/help)

If you need help using Doh, or have found a bug, please create an issue on the
<a href="https://github.com/pjfl/p5-app-doh/issues" target="_blank">Github
repo</a>.

Please note that a Perl module failing to install is not an issue for *this*
application. Each Perl module has an issues tracker which can be found via that
modules [Meta::CPAN](https://metacpan.org) page


Changes
Revision history for App-Doh

          - Fixed renaming posts
          - Fixed configuration assetdir attribute default value
          - Fixed daemon access log and pid file names

0.15.1    2015-12-18 16:25:27
          - Merged docs path config into main file
          - Moved users.json to ctrldir
          - Switch log and po filenames to class2appdir appclass
          - Randomised secret generation

0.14.1    2015-12-09 01:28:40
          - Default mount point /doh allows for redirect to static index
          - Fixed Config components needed to be lazy
          - Made env_var into a class method
          - Uses Web::Components::Role::TT
          - Idiomatic rename of Functions to Util
          - Moving Role::Template to own distribution

0.13.1    2015-08-31 18:23:12
          - Updated prereqs

0.12.1    2015-08-31 18:12:16
          - Merged Config common_links into stash_attr
          - Fixed locale/language leaky abstraction from request object
          - Added expected_rv parameters to Daemon
          - Merged Preference and CommonLinks roles into PageConfiguration
          - Moved Usul attribute constructor to Functions

0.11.1    2015-08-28 12:54:54
          - Created Config stash_attr from preferences and lists in code
          - Passing tests again
          - Switched to Web::Components
          - Enhanced type constraint names
          - Renamed Web::Components loader and role
          - Made APP_DOH the env var prefix for consistancy
          - Stopped putting request object in stash
          - Made cdnjs configurable via jslibs in Config
          - Stopped using state keyword
          - Model execute now throws if method not found
          - Added support for making Debian packages
          - Added missing middleware dependecies
          - Simplified load_components signature
          - Added make-site-tarball
          - Added caching to generate make function
          - Config monikers was replace with components
          - Model get_stash was renamed to initialise_stash
          - Updated prereqs
          - Made program wrappers work in taint mode

0.10.1    2015-05-11 00:39:33
          - Fixed memory leak in component DI
          - Made User class a model component and used DI
          - Implement dependency injection in components
          - ComponentLoading returns plain text 500 if Request->new fails
          - Added post_install and uninstall commands to CLI
          - Scripts sources a local env script
          - Toolchain now builds a Debian file

0.9.1     2015-01-04 16:01:57
          - Ajax requests now use JSON not XML
          - Added model_name to request object
          - API improvement ::Functions::load_components
          - Fixed search methods external call to ack
          - Made private functions and methods lexical subs

0.8.1     2014-10-31 01:08:36
          - Updated throw call syntax
          - Fixed decoding of body parameters
          - Component and input encoding defaults to config
          - Added maximum generation run time
          - URIs now properly enscaped
          - Improved Request API will instantiate from a hash

0.7.1     2014-09-19 00:23:42
          - Improved Request API raw params now private
          - Session secret now has more source options
          - Component loading respects + at the start of class names
          - Static generation no longer creates defunct process
          - Added editor CSS to repo for install off Github
          - Enabled ConditionalGET in Plack stack
          - Serve JS from content delivery network
          - Added t::boilerplate to tests

0.6.1     2014-08-07 15:39:42
          - Added Data::Validation to user form
          - Implemented status message queue
          - Get /favicon.ico now returns 404
          - Added make_skin method to CLI
          - Renamed less files to create skin structure
          - Improved layout of profile-user form
          - Put page status on a div since defaultStatus not supported
          - Removed ForwardEvents class from JS. Doesn't work
          - Created editor.tt for the markdown editor
          - Style fixes to post navigation buttons

0.5.1     2014-07-29 18:35:58
          - Made Root controller look like a routing table
          - Made component loading a role
          - Added keybindings for the markdown editor
          - Refactored markdown editor
          - Switched status messages to JS notice board

0.4.1     2014-07-26 01:04:29
          - Added insertGraves to markdown editor
          - Added CodeMirror JS markdown editor
          - Added config user attribute to own Daemon process
          - Added Role::Component. Doh.pm contains only VERSION number
          - Streamlined the controller code
          - Server throws better exceptions
          - Patched bug in warnings::illegalproto for Perl < 5.012
          - Pass appclass in from doh-server
          - Abstracted apps env var access out to a function
          - Split Behaviour js class out into own file
          - Markdown allows class on trailing three graves
          - Added default view configuration option
          - Use App::Doh in ::Server to get VERSION in scope

0.3.1     2014-07-19 20:44:34
          - Tested on Chrome and Opera
          - Server now loads MVC components at runtime
          - Rationalised the templates
          - Fixed footer and add JS handler for content margin-bottom
          - Added minimum password length
          - Admin page functional improvements
          - Improved CSS for forms
          - Upated metadata
          - Added user profile management
          - Added ::Session to MANIFEST
          - Added ::Session. Added session timeout
          - Adjusted CSS tips box model
          - Adjusted CSS prompt widths
          - Removed unsed dependency
          - Fixed tests
          - Admin interface update user roles

lib/Doh/Config.pm
# @(#)Ident: Config.pm 2013-08-23 20:05 pjf ;

package Doh::Config;

use namespace::sweep;
use version; our $VERSION = qv( sprintf '0.1.%d', q$Rev: 19 $ =~ /\d+/gmx );

use Class::Usul::Constants;
use File::DataClass::Types  qw( ArrayRef Directory HashRef NonEmptySimpleStr
                                NonNumericSimpleStr NonZeroPositiveInt
                                SimpleStr );
use Moo;

extends q(Class::Usul::Config::Programs);

has 'author'           => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'Dave';

has 'brand'            => is => 'ro',   isa => SimpleStr, default => NUL;

has 'colours'          => is => 'lazy', isa => ArrayRef, init_arg => undef;

has 'common_links'     => is => 'ro',   isa => ArrayRef,
   default             => sub { [ qw( css help_url images less js ) ] };

has 'css'              => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'css/';

has 'description'      => is => 'ro',   isa => SimpleStr, default => NUL;

has 'docs_path'        => is => 'lazy', isa => Directory,
   coerce              => Directory->coercion,
   default             => sub { $_[ 0 ]->root->catdir( 'docs' ) };

has 'float'            => is => 'ro',   isa => NonNumericSimpleStr,
   coerce              => sub { $_[ 0 ] ? 'float-view' : NUL }, default => TRUE;

has 'font'             => is => 'ro',   isa => SimpleStr,
   default             => sub {
      'http://fonts.googleapis.com/css?family=Roboto+Slab:400,700,300,100' };

has 'google_analytics' => is => 'ro',   isa => SimpleStr, default => NUL;

has 'help_url'         => is => 'ro',   isa => SimpleStr, default => 'help';

has 'images'           => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'img/';

has 'js'               => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'js/';

has 'keywords'         => is => 'ro',   isa => SimpleStr, default => NUL;

has 'less'             => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'less/';

has 'links'            => is => 'lazy', isa => ArrayRef, init_arg => undef;

has 'mount_point'      => is => 'ro',   isa => NonEmptySimpleStr,
   default             => '/';

has 'no_index'         => is => 'ro',   isa => ArrayRef,
   default             => sub { [ qw( .git .svn cgi-bin doh.json ) ] };

has 'port'             => is => 'lazy', isa => NonZeroPositiveInt,
   default             => 8085;

has 'preferences'      => is => 'ro',   isa => ArrayRef,
   default             => sub { [ qw( float skin theme ) ] };

has 'projects'         => is => 'ro',   isa => HashRef,   default => sub { {} };

has 'repo_url'         => is => 'ro',   isa => SimpleStr, default => NUL;

has 'server'           => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'Twiggy';

has 'skin'             => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'default';

has 'template'         => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'index';

has 'theme'            => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'green';

has 'title'            => is => 'ro',   isa => NonEmptySimpleStr,
   default             => 'Documentation';

has 'twitter'          => is => 'ro',   isa => ArrayRef,  default => sub { [] };


has '_colours'         => is => 'ro',   isa => HashRef,
   default             => sub { {} }, init_arg => 'colours';

has '_links'           => is => 'ro',   isa => HashRef,
   default             => sub { {} }, init_arg => 'links';

sub _build_colours {
   return __to_array_of_hash( $_[ 0 ]->_colours, qw( key value ) );
}

sub _build_links {
   return __to_array_of_hash( $_[ 0 ]->_links, qw( name url ) );
}

sub __to_array_of_hash {
   my ($href, $key_key, $val_key) = @_;

   return [ map { my $v = $href->{ $_ }; +{ $key_key => $_, $val_key => $v } }
            sort keys %{ $href } ],
}

1;

__END__

=pod

=encoding utf8

=head1 Name

Doh::Config - One-line description of the modules purpose

=head1 Synopsis

   use Class::Usul;

   Class::Usul->new( config_class => 'Doh::Config' );

=head1 Version

This documents version v0.1.$Rev: 19 $ of L<Doh::Config>

=head1 Description

=head1 Configuration and Environment

Defines the following attributes;

=over 3

=item C<author>

A non empty simple string that defaults to 'Dave'

=item C<brand>

A simple string that defaults to null

=item C<colours>

A lazily evaluated array ref

=item C<common_links>

An array ref that defaults to '[ qw( css help_url images less js ) ]'

=item C<css>

A non empty simple string that defaults to 'css/'

=item C<description>

A simple string that defaults to null

=item C<docs_path>

A lazily evaluated directory that defaults to 'var/root/docs'

=item C<float>

A non numeric simple string which defaults to 'float-view'

=item C<font>

A simple string that defaults to
'http://fonts.googleapis.com/css?family=Roboto+Slab:400,700,300,100'

=item C<google_analytics>

A simple string that defaults to null

=item C<help_url>

A simple string that defaults to 'help'

=item C<images>

A non empty simple string that defaults to 'img/'

=item C<js>

A non empty simple string that defaults to 'js/'

=item C<keywords>

A simple string that defaults to null

=item C<less>

A non empty simple string that defaults to 'less/'

=item C<links>

A lazily evaluated array ref

=item C<mount_point>

A non empty simple string that defaults to '/'

=item C<no_index>

An array ref that defaults to '[ qw( .git .svn cgi-bin doh.json ) ]'

=item C<port>

A lazily evaluated non zero positive integer that defaults to 8085

=item C<preferences>

An array ref that defaults to '[ qw( float skin theme ) ]'

=item C<projects>

A hash ref that defaults to an empty hash ref

=item C<repo_url>

A simple string that defaults to null

=item C<server>

A non empty simple string that defaults to 'Twiggy'

=item C<skin>

A non empty simple string that defaults to 'default'

=item C<template>

A non empty simple string that defaults to 'index'

=item C<theme>

A non empty simple string that defaults to 'green'

=item C<title>

A non empty simple string that defaults to 'Documentation'

=item C<twitter>

An array ref that defaults to an empty array ref

=back

=head1 Subroutines/Methods

None

=head1 Diagnostics

None

=head1 Dependencies

=over 3

=item L<Class::Usul>

=item L<File::DataClass>

=item L<Moo>

=back

=head1 Incompatibilities

There are no known incompatibilities in this module

=head1 Bugs and Limitations

There are no known bugs in this module.
Please report problems to the address below.
Patches are welcome

=head1 Acknowledgements

Larry Wall - For the Perl programming language

=head1 Author

Peter Flanigan, C<< <pjfl@cpan.org> >>

=head1 License and Copyright

Copyright (c) 2013 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

=cut

# Local Variables:
# mode: perl
# tab-width: 3
# End:
